### PR TITLE
Login: Magic Login: Support two step / two factor authentication & improve user-facing flow

### PR DESCRIPTION
**Pull request by jblz**
Adds a route to handle clickthroughs: `/log-in/link/use/:lang?` along with styling improvements.

## To Test:
Apply D5969-code & complete that test plan

### CODE REVIEW

File: client/state/login/magic-login/actions.js [hunk 1]
```diff
@@ -93,39 +88,36 @@ const authSuccess = () => {
 	};
 };
```
Comment (stephanethomas) [inline client/state/login/magic-login/actions.js]: Do we really need to have those `authError()` and `authSuccess()` functions ? As far I can see they are only used once in this file.

File: client/login/controller.js [hunk 2]
```diff
@@ -18,8 +19,22 @@ export default {
...
 	},
+
 	magicLogin( context, next ) {
 		context.primary = <MagicLogin />;
 		next();
-	}
+	},
+
+	magicLoginUse( context, next ) {
+		// queryArguments isn't set in redux in time for this initial render -- pull 'em out here.
```
Comment (mcsf) [inline client/login/controller.js]: Yep, controllers are good at this, feel free to remove the comment altogether!

File: client/login/controller.js [hunk 3]
```diff
@@ -18,8 +19,22 @@ export default {
...
+		const {
+			client_id,
+			email,
+			token,
+			tt,
+		} = context.query;
+
+		context.primary = <HandleEmailedLinkForm clientId={ client_id } emailAddress={ email } token={ token } tokenTime={ tt } />;
```
Comment (mcsf) [inline client/login/controller.js]: I suggest breaking the line, but personal pref.

File: client/login/magic-login/handle-emailed-link-form.jsx [hunk 4]
```diff
@@ -40,42 +76,75 @@ class HandleEmailedLinkForm extends React.Component {
...
+	rebootAfterLogin = () => {
+		const { redirectTo } = this.props;
+
+		this.props.recordTracksEvent( 'calypso_login_success', {
+			two_factor_enabled: this.props.twoFactorEnabled,
+			magic_login: 1
+		} );
+
+		// Redirects to / if no redirect url is available
+		const url = redirectTo ? redirectTo : window.location.origin;
```
Comment (mcsf) [inline client/login/magic-login/handle-emailed-link-form.jsx]: Perhaps a dumb question: what is `window.location.origin` giving us that beats `'/'`? Could there be exceptions to bite us? Or inconsistent implementations across platforms? (This was prompted by the fact that a component's reliance on `window` makes it less flexible for testing, server-side rendering, etc.)

File: client/login/magic-login/handle-emailed-link-form.jsx [hunk 5]
```diff
@@ -40,42 +76,75 @@ class HandleEmailedLinkForm extends React.Component {
...
+
+		// user data is persisted in localstorage at `lib/user/user` line 157
+		// therefor we need to reset it before we redirect, otherwise we'll get
```
Comment (mcsf) [inline client/login/magic-login/handle-emailed-link-form.jsx]: Typo at `therefor`, my liege. _/me bows ever so humbly_

File: client/login/magic-login/handle-emailed-link-form.jsx [hunk 6]
```diff
@@ -40,42 +76,75 @@ class HandleEmailedLinkForm extends React.Component {
...
+		// mixed data from old and new user
+		user.clear();
+
+		window.location.href = url;
```
Comment (mcsf) [inline client/login/magic-login/handle-emailed-link-form.jsx]: Why not `page()`?

File: client/login/magic-login/handle-emailed-link-form.jsx [hunk 7]
```diff
@@ -106,10 +175,13 @@ class HandleEmailedLinkForm extends React.Component {
 				</p> );
 		}
 
+		this.props.recordTracksEvent( 'calypso_login_magic_link_handle_click_view' );
+
 		return (
 			<EmptyContent
 				action={ action }
 				className={ classNames( {
+					'magic-login__handle-link': true,
```
Comment (mcsf) [inline client/login/magic-login/handle-emailed-link-form.jsx]: `classNames` supports the following:

```js
className={ classNames( 'magic-login__handle-link', 'some-other-optional-class', {
  'magic-login__is-fetching-auth': isFetching,
} ) }
```

File: client/login/magic-login/index.jsx [hunk 8]
```diff
@@ -56,78 +50,66 @@ class MagicLogin extends React.Component {
...
-		} = this.props;
+	pageViewTracker() {
+		const path = '/log-in/link';
+		const title = 'Login > Link';
 
-		if ( magicLoginEnabled && queryArguments && queryArguments.action === 'handleLoginEmail' ) {
-			this.props.showMagicLoginInterstitialPage();
-		}
+		return <PageViewTracker path={ path } title={ title } />;
 	}
```
Comment (mcsf) [inline client/login/magic-login/index.jsx]: Came to say the same. 😄

File: client/login/magic-login/request-login-email-form.jsx [hunk 9]
```diff
@@ -78,19 +76,19 @@ class RequestLoginEmailForm extends React.Component {
 				<h1 className="magic-login__form-header">
 					{ translate( 'Email me a login link.' ) }
 				</h1>
-				<LoggedOutForm onSubmit={ this.onSubmit }>
+				<p>{
+					translate( 'Get a link sent to the email address associated ' +
```
Comment (mcsf) [inline client/login/magic-login/request-login-email-form.jsx]: Nit-picky, but style is more usually:

```js
<p>
    { translate( …
        … ) }
</p>
```

and sometimes with the closing curly on its own line, but personally I think that adds noise.

File: client/login/magic-login/handle-emailed-link-form.jsx [hunk 10]
```diff
@@ -40,42 +76,75 @@ class HandleEmailedLinkForm extends React.Component {
...
+		if ( ! this.props.twoFactorEnabled ) {
+			this.rebootAfterLogin();
+		} else {
+			page( login( {
+				isNative: true,
+				// If no notification is sent, the user is using the authenticator for 2FA by default
+				twoFactorAuthType: this.props.twoFactorNotificationSent.replace( 'none', 'authenticator' )
+			} ) );
 		}
+	};
```
Comment (mcsf) [inline client/login/magic-login/handle-emailed-link-form.jsx]: Handlers like this one only depend on Redux-connected props — not own props — for their logic. To me, that further gives us reason to treat this as an action and move it out of the component. Same goes for `rebootAfterLogin` — actually, they're intertwined.

File: client/login/magic-login/handle-emailed-link-form.jsx [hunk 11]
```diff
@@ -40,42 +76,75 @@ class HandleEmailedLinkForm extends React.Component {
...
-		this.props.showMagicLoginLinkExpiredPage();
-	}
+	// Lifted from `blocks/login`
+	// @TODO unify
+	rebootAfterLogin = () => {
+		const { redirectTo } = this.props;
+
+		this.props.recordTracksEvent( 'calypso_login_success', {
+			two_factor_enabled: this.props.twoFactorEnabled,
+			magic_login: 1
```
Comment (mcsf) [inline client/login/magic-login/handle-emailed-link-form.jsx]: As https://github.com/Automattic/eslint-config-wpcalypso/pull/5 should be merged soon, I recommend adding a trailing comma there.

File: client/login/magic-login/handle-emailed-link-form.jsx [hunk 12]
```diff
@@ -1,36 +1,72 @@
...
+		isExpired: PropTypes.bool,
+		isFetching: PropTypes.bool,
+		twoFactorEnabled: PropTypes.bool,
+		twoFactorNotificationSent: PropTypes.string,
+
+		// Conntected action creators
+		fetchMagicLoginAuthenticate: PropTypes.func.isRequired,
+		recordTracksEvent: PropTypes.func.isRequired,
+		showMagicLoginLinkExpiredPage: PropTypes.func.isRequired,
+	};
```
Comment (mcsf) [inline client/login/magic-login/handle-emailed-link-form.jsx]: I agree with Stéphane: I'd like to see more decoupled and atomic pieces here.

File: client/state/login/magic-login/actions.js [hunk 13]
```diff
@@ -72,66 +66,48 @@ export const fetchMagicLoginRequestEmail = email => dispatch => {
...
-	};
-};
-
-const authSuccess = () => {
-	return {
-		type: MAGIC_LOGIN_REQUEST_AUTH_SUCCESS,
-	};
-};
-
+// @TODO should this move to `/state/data-layer`..?
```
Comment (jblz) [inline client/state/login/magic-login/actions.js]: Any opinions here? or is this ok where it is?

File: client/state/login/magic-login/actions.js [hunk 14]
```diff
@@ -72,66 +66,48 @@ export const fetchMagicLoginRequestEmail = email => dispatch => {
...
+			client_secret: config( 'wpcom_signup_key' ),
+			email,
+			token,
+			tt,
+		} )
+		.then( ( response ) => {
+			dispatch( {
+				type: LOGIN_REQUEST_SUCCESS,
+				data: get( response, 'body.data' ),
+				// @TODO figure how we should treat `rememberMe`...
```
Comment (jblz) [inline client/state/login/magic-login/actions.js]: Any opinions here? I'm not 💯 on the implications of setting `rememberMe` here -- does it just affect the TTL on the cookie?

File: client/login/magic-login/handle-emailed-link-form.jsx (hunk 6 above)
Comment (stephanethomas) [inline client/login/magic-login/handle-emailed-link-form.jsx]: That's because we want to reload Calypso, and not simply redirect the user to the home page.

File: client/login/magic-login/handle-emailed-link-form.jsx (hunk 4 above)
Comment (stephanethomas) [inline client/login/magic-login/handle-emailed-link-form.jsx]: There is no dumb question :).

However, given the fact that we're using `window.location.href` below I'm afraid this component will still rely on `window`.

File: client/state/login/magic-login/actions.js (hunk 13 above)
Comment (stephanethomas) [inline client/state/login/magic-login/actions.js]: Reading the [readme file](https://github.com/Automattic/wp-calypso/blob/master/client/state/data-layer/README.md) it looks like it's where we should put this type of code now. It's probably better if we move all login actions in a different pull request though.

File: client/state/login/magic-login/actions.js (hunk 14 above)
Comment (stephanethomas) [inline client/state/login/magic-login/actions.js]: Yes, it only affects the lifespan of the cookie.

If the default value is fine for magic login, we could probably update [the reducer](https://github.com/Automattic/wp-calypso/blob/a82dd1af35ec0b211f2918326fb05e1e7e5f558c/client/state/login/reducer.js#L49-L53) so that you don't have to specify `rememberMe` here (and also fix the state returned for the `LOGIN_REQUEST_FAILURE` event that should be `null` instead of `false`):

```
export const rememberMe = createReducer( null, {
  [ LOGIN_REQUEST ]: () => null,
  [ LOGIN_REQUEST_SUCCESS ]: ( state, action ) => get( data, 'rememberMe', false ),
  [ LOGIN_REQUEST_FAILURE ]: () => null,
} );
```

File: client/login/magic-login/emailed-login-link-expired.jsx [hunk 15]
```diff
@@ -12,14 +12,23 @@ import addQueryArgs from 'lib/route/add-query-args';
...
+	static propTypes = {
+		recordPageView: PropTypes.func.isRequired,
+		translate: PropTypes.func.isRequired,
+	};
+
 	render() {
-		const { translate } = this.props;
+		const {
+			translate,
+		} = this.props;
```
Comment (stephanethomas) [inline client/login/magic-login/emailed-login-link-expired.jsx]: This is absolutely nitpicking but I don't see a reason for breaking that destructuring into three lines.

File: client/login/magic-login/index.jsx [hunk 16]
```diff
@@ -4,130 +4,88 @@
...
 		return (
-			<Main className={ {
-				'magic-login': true,
-				'magic-login__request_link': ! magicLoginView || magicLoginView === REQUEST_FORM,
-			} }>
-				<PageViewTracker path="/log-in/link" title="Login > Link" />
-
+			<Main className={ classNames( 'magic-login', {
+				'magic-login__request_link': ! showCheckYourEmail,
+			} ) }>
```
Comment (stephanethomas) [inline client/login/magic-login/index.jsx]: We could make this block a little bit easier to read by moving this `classNames()` call above in a variable.

File: client/login/magic-login/request-login-email-form.jsx [hunk 17]
```diff
@@ -10,29 +10,50 @@ import { connect } from 'react-redux';
...
+	};
+
 	onEmailAddressFieldChange = event => {
-		this.props.setMagicLoginInputEmailAddress( event.target.value );
+		this.setState( {
+			emailAddress: event.target.value,
+		} );
 	};
 
 	onNoticeDismiss = () => {
```
Comment (stephanethomas) [inline client/login/magic-login/request-login-email-form.jsx]: Not related to this pull request, but this block should end with a semicolon:

```
onNoticeDismiss = () => {
  this.props.hideMagicLoginRequestNotice();
};
```

File: client/login/magic-login/request-login-email-form.jsx [hunk 18]
```diff
@@ -10,29 +10,50 @@ import { connect } from 'react-redux';
...
 import FormTextInput from 'components/forms/form-text-input';
 import LoggedOutForm from 'components/logged-out-form';
 import Notice from 'components/notice';
 import { localize } from 'i18n-calypso';
-
-//@TODO move this to wp-login compoenent
 import { getCurrentUser } from 'state/current-user/selectors';
 
 class RequestLoginEmailForm extends React.Component {
+	state = {};
```
Comment (stephanethomas) [inline client/login/magic-login/request-login-email-form.jsx]: Just curious, do we really need to declare the state like this if it is empty?

File: client/login/magic-login/request-login-email-form.jsx [hunk 19]
```diff
@@ -41,19 +62,27 @@ class RequestLoginEmailForm extends React.Component {
...
 			requestError,
 			isFetching,
-			isValidEmailAddress,
 			emailRequested,
+			showCheckYourEmail,
 			translate,
 		} = this.props;
 
+		if ( showCheckYourEmail ) {
+			this.props.recordPageView( '/log-in/link', 'Login > Link > Emailed' );
```
Comment (stephanethomas) [inline client/login/magic-login/request-login-email-form.jsx]: Shouldn't we move this call inside `<EmailedLoginLinkSuccessfully />` instead (similarly to what was done with `<EmailedLoginLinkExpired />`?

File: client/login/magic-login/request-login-email-form.jsx (hunk 18 above)
Comment (jblz) [inline client/login/magic-login/request-login-email-form.jsx]: @stephanethomas -- it was to prevent a reference error in the render method before any state was set. I went ahead and took it out & put a guard in place in ff01d10 which should be even more robust.

**Comment by stephanethomas** [comment]
I reviewed the server patch, and tested those changes. Everything seems to work fine.

I'm not sure if it's possible, but it would be great to remove the following screen from the flow as it doesn't bring any value to the user:

screenshot
